pkg/controller/schedulingmanager/controller.go

```diff
@@ -0,0 +1,176 @@
+
+	c.store, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			// Only watch the federation namespace to ensure
+			// restrictive authz can be applied to a namespaced
+			// control plane.
+			ListFunc: func(options metav1.ListOptions) (pkgruntime.Object, error) {
+				return client.FederatedTypeConfigs(fedNamespace).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return client.FederatedTypeConfigs(fedNamespace).Watch(options)
+			},
+		},
+		&corev1a1.FederatedTypeConfig{},
+		util.NoResyncPeriod,
+		util.NewTriggerOnAllChanges(c.worker.EnqueueObject),
+	)
+
+	return c, nil
+}
+
+// Run runs the Controller.
+func (c *SchedulerController) Run(stopChan <-chan struct{}) {
+	go c.controller.Run(stopChan)
+
+	// wait for the caches to synchronize before starting the worker
+	if !cache.WaitForCacheSync(stopChan, c.controller.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for cache to sync"))
+		return
+	}
+
+	c.worker.Run(stopChan)
+}
+
+func (c *SchedulerController) reconcile(qualifiedName util.QualifiedName) util.ReconciliationStatus {
+	key := qualifiedName.String()
+
+	glog.Infof("Running reconcile FederatedTypeConfig for %q", key)
+
+	cachedObj, exist, err := c.store.GetByKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to query FederatedTypeConfig store for %q: %v", key, err))
+		return util.StatusError
+	}
+	if !exist {
+		return util.StatusAllOK
+	}
+	typeConfig := cachedObj.(*corev1a1.FederatedTypeConfig)
+
+	deleted := typeConfig.DeletionTimestamp != nil
+	if deleted {
+		// Do not support deletion yet
+		return util.StatusAllOK
+	}
+
+	// set name and group for the type config target
+	corev1a1.SetFederatedTypeConfigDefaults(typeConfig)
+	apiResource := typeConfig.GetTarget()
+	templateKind := typeConfig.GetTemplate().Kind
+
+	for kind, schedulingType := range schedulingtypes.SchedulingTypes() {
+		for _, fedKind := range SchedulingRegistry[kind] {
+			if fedKind == templateKind {
+				// Scheduling preference controller is started on demand
+				_, ok := c.scheduler[kind]
+				if !ok {
+					scheduler, err := schedulingpreference.StartSchedulingPreferenceController(kind, schedulingType.SchedulerFactory, c.config, c.fedNamespace, c.clusterNamespace, c.targetNamespace, c.stopChan, util.DefaultClusterAvailableDelay, util.DefaultClusterUnavailableDelay, c.minimizeLatency)
+					if err != nil {
+						runtime.HandleError(fmt.Errorf("Error starting schedulingpreference controller for %q : %v", kind, err))
+						return util.StatusError
+					}
+					c.scheduler[kind] = scheduler
+				}
+
+				glog.Infof("Start plugin with kind %s for scheduling type %s", templateKind, kind)
+				err := c.scheduler[kind].StartPlugin(templateKind, &apiResource, c.stopChan)
+				if err != nil {
+					runtime.HandleError(fmt.Errorf("Error starting plugin for %q : %v", templateKind, err))
+					return util.StatusError
+				}
+				break
```

Comment:
Why is it desirable to break out of this loop but not the outer one?

Reply:
Currently, we have "ReplicaSchedulingPreference" scheduler.  schedulingtypes.RegisterSchedulingType() support multiple schedulers. 

If a federated resource can be registered into more than 1 scheduler, break out of inner loop should be fine. If the resource cannot be registered into multiple scheduler (may introduce any conflict), it will not be defined in our SchedulingRegistry. It is also fine break from inner loop.

Of course, if we can sure there is no multiple registration case, break out of outside loop is fine.

Reply:
I value consistency, and I think either both loops should be broken out of or none should to avoid confusing future maintainers.